### ISSUE TITLE
[Vertical Pod Autoscaler Metrics] Add Support for new "vs" Requests / Limits Metrics

### ISSUE THREAD

**Issue by sherifkayad** [kubernetes/kube-state-metrics#1536]
**What would you like to be added**:
I would like to see the following new metrics added to the list of VPA metrics:
- kube_verticalpodautoscaler_status_recommendation_containerrecommendations_$RECOMMENDATION_vs_requests e.g. `kube_verticalpodautoscaler_status_recommendation_containerrecommendations_target_vs_requests` => to indicate the status of the recommendation vs. the allocated resource requests e.g. a value >0, =0 or <0 to indicate if the recommendation is greater than, equals or is less than the allocated resource requests (one could also think about the ratio between the recommendation vs. the requests in order to see how big the deviation is)
- kube_verticalpodautoscaler_status_recommendation_containerrecommendations_$RECOMMENDATION_vs_limits e.g. `kube_verticalpodautoscaler_status_recommendation_containerrecommendations_target_vs_limits` => to indicate the status of the recommendation vs. the allocated resource limits e.g. a value >0, =0 or <0 to indicate if the recommendation is greater than, equals or is less than the allocated resource limits (one could also think about the ratio between the recommendation vs. the limit in order to see how big the deviation is)

**Why is this needed**:
Basically we have an alerting use case to alert our teams whenever there's a (big) deviation between the recommendation values and the currently set requests and limits so that the teams can take action. Here we don't really want to use the VPA in an `update` mode to avoid restarts at critical times and since we have a long monitoring cycle before taking actions.

**Describe the solution you'd like**
As mentioned in the "What would you like to be added", here the proposed solution is adding two new metrics for each recommendation type (total of 8 new metrics) to indicate the deviation between the recommendation and the resource requests / limits for cpu and memory.

**Additional context**
Nothing further to add

**Comment by fpetkovski** [comment]
Hi @sherifkayad, I think the use case you're describing is legitimate and useful. I was just wondering if the same can be achieved using a promQL query using the existing metrics that are currently provided by KSM.

**Comment by sherifkayad** [comment]
@fpetkovski Thanks a lot for your prompt reply .. After tons of digging I found out that the "only" way to achieve that using existing queries will be something like that:
```
(sum(kube_verticalpodautoscaler_status_recommendation_containerrecommendations_target{resource="cpu",target_name="my-app"}) by (namespace,target_name,container)) /

on (container) group_right(target_name)

(sum(kube_pod_container_resource_requests{resource="cpu",pod=~"my-app(.+)"}) by (container,pod))

* 100
```
**OR**
```
(sum(kube_verticalpodautoscaler_status_recommendation_containerrecommendations_target{resource="cpu",target_name="my-app"}) by (namespace,target_name,container)) /

on (container) group_right(target_name)

(sum(kube_pod_container_resource_limits{resource="cpu",pod=~"my-app(.+)"}) by (container,pod))

* 100
```

It's a fact that the query above will get you the percentage of the recommendation vs. the requests or limits. However, there are the following limitations:
- I have to explicitly state the `target_name` as well as the `resource` in my query; otherwise the PromQL will fail .. take into consideration the use case when you have multiple pods e.g. replica set with x instances where x > 1
- Doing the workaround above will result in an alert per VPA target! .. Ideally I want to have two alerts for all apps (one for the requests and one for the limits in comparison with e.g. the VPA `target` value)

**Comment by fpetkovski** [comment]
With regards to adding the additional metric, this is not easily achievable with the current design of KSM. I am also not sure if it is something we want to support. The reason for this is because each metric is related to exactly one resource. Adding the metrics which are proposed here would mean that we need to start correlating labels and metrics across resources. 

From what I understand, the problem you're running into is the fact that a VPA works on deployments, while resources are set on individual pods/containers. I would suggest looking into a query similar to the following one, so that you can identify each pod's governing deployment:
```
label_replace(
  label_replace(
    kube_pod_owner{namespace="default"}, "replicaset", "$1", "owner_name", "(.*)"
  ) * on (replicaset) group_left(owner_name) kube_replicaset_owner{namespace="default"}, 
  "target_name", "$1", "owner_name", "(.*)"
) 
* on (target_name) group_left(container) kube_verticalpodautoscaler_status_recommendation_containerrecommendations_target{resource="cpu"}
-  on (pod, container) kube_pod_container_resource_requests{resource="cpu", namespace="default"}
```

It's a loaded expression, but the gist of it is that it uses `kube_pod_owner` and `kube_replicaset_owner` to get the link between pods and deployments.

**Comment by sherifkayad** [comment]
@fpetkovski how about StatefulSet's? .. Is it possible to have both deployments and Statefulset's included in the search at the same time?

**Comment by fpetkovski** [comment]
Statefulsets manage pods directly, so you don't need the `kube_replicaset_owner` metric:
```
label_replace(kube_pod_owner{namespace="default"}, "target_name", "$1", "owner_name", "(.*)") 
* on (target_name) group_left(container) kube_verticalpodautoscaler_status_recommendation_containerrecommendations_target{resource="cpu"}
-  on (pod, container) kube_pod_container_resource_requests{resource="cpu", namespace="default"}
```

You can try merging those two in a single expression, or keep them as two different alerts.

**Comment by sherifkayad** [comment]
Amazing! These are my queries for those interested (scanning all namespaces):

```
## DEPLOYMENT::
label_replace(
  label_replace(
    kube_pod_owner, "replicaset", "$1", "owner_name", "(.*)"
  ) * on (replicaset, namespace) group_left(owner_name) kube_replicaset_owner, 
  "target_name", "$1", "owner_name", "(.*)"
) 
* on (target_name, namespace) group_left(container) kube_verticalpodautoscaler_status_recommendation_containerrecommendations_target{resource="cpu"}
-  on (pod, container, namespace) group_right(target_name) kube_pod_container_resource_requests{resource="cpu"}

## STATEFULSET::
label_replace(kube_pod_owner, "target_name", "$1", "owner_name", "(.*)") 
* on (target_name, namespace) group_left(container) kube_verticalpodautoscaler_status_recommendation_containerrecommendations_target{resource="cpu"}
-  on (pod, container, namespace) group_right(target_name) kube_pod_container_resource_requests{resource="cpu"}
```

**Comment by sherifkayad** [comment]
@fpetkovski seems like I am facing an issue with pods that have multiple containers. Is there a workaround for that? .. I can't seem to find the right PromQL query to solve that
